src/models/state/mempool.rs

```diff
@@ -303,12 +314,28 @@ impl Mempool {
     /// Panics if the transaction's proof is of the wrong type.
     pub(super) fn insert(
         &mut self,
-        transaction: Transaction,
+        new_tx: Transaction,
         origin: TransactionOrigin,
     ) -> Vec<MempoolEvent> {
+        fn new_tx_has_higher_proof_quality(
+            new_tx: &Transaction,
+            conflicts: &[(TransactionKernelId, &Transaction)],
+        ) -> bool {
+            match &new_tx.proof {
+                TransactionProof::Invalid => panic!("Invalid proofs don't belong in mempool"),
+                TransactionProof::Witness(_) => false,
+                TransactionProof::ProofCollection(_) => conflicts
+                    .iter()
+                    .any(|x| matches!(&x.1.proof, TransactionProof::Witness(_))),
+                TransactionProof::SingleProof(_) => conflicts
+                    .iter()
+                    .any(|x| !matches!(&x.1.proof, TransactionProof::SingleProof(_))),
+            }
+        }
+
         let mut events = vec![];
 
-        match transaction.proof {
+        match new_tx.proof {
             TransactionProof::Invalid => panic!("cannot insert invalid transaction into mempool"),
```

Comment:
same nit as above.  If Mempool only accepts a type that doesn't include Invalid variant, then caller has to deal with the issue somehow, and its not mempool's problem.  Probably it should be dealt with when its first determined the proof is invalid.